### ISSUE TITLE
doubt about the input in the bi-lstm model

### ISSUE THREAD

**Issue by OswaldoBornemann** [dongjun-Lee/birnn-language-model-tf#3]
The input and output code is below

```
self.lm_input = self.x
self.lm_output = self.x[:, 1:-1]
```
I ran a test, 

```
test[:, :][0] = array([   1,  101,    3, 1105,   15,  247,    4,   30,  285,    2,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0])

test[:, 1:-1][0] = array([ 101,    3, 1105,   15,  247,    4,   30,  285,    2,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0])
```

But i looked up the paper, shouldn't the real input and output is like below ?

```
input = [1,  101,    3, 1105,   15,  247,    4,   30,  285 ] 
output = [101,    3, 1105,   15,  247,    4,   30,  285,    2]
```

**Comment by dongjun-Lee** [comment]
Hi, paddings(0) will be ignored by ```sequence_length``` at [this line](https://github.com/dongjun-Lee/birnn-language-model-tf/blob/master/model/bi_rnn_lm.py#L33).

**Comment by OswaldoBornemann** [comment]
yes, i know that. My question is that the real input should not be `data[:-1]` and the target is `data[1:]` ?

**Comment by OswaldoBornemann** [comment]
I may not state that clearly. In your example, the real input is `[1,  101,    3, 1105,   15,  247,    4,   30,  285,    2]` and real output is `[ 101,    3, 1105,   15,  247,    4,   30,  285,    2]`. But what i think is that the real input should be `[1,  101,    3, 1105,   15,  247,    4,   30,  285]` and the real output is `[ 101,    3, 1105,   15,  247,    4,   30,  285,    2]`

**Comment by dongjun-Lee** [comment]
Do you mean the last token(2) in the input sequence should be removed? It will be also ignored by sequence_length.

**Comment by dongjun-Lee** [comment]
You mean the last token in the input(2) should be removed?

**Comment by OswaldoBornemann** [comment]
@dongjun-Lee yes, that's what i mean. How do you think? Can you explain ?

**Comment by OswaldoBornemann** [comment]
@dongjun-Lee This problem makes me suffered... I even want to call you....

**Comment by dongjun-Lee** [comment]
@tsungruihon I think the last token (```<eos>```) in the ```self.lm_input``` should not be removed because of the backward RNN. At [here](https://github.com/dongjun-Lee/birnn-language-model-tf/blob/master/model/bi_rnn_lm.py#L32-L33), we encode full sentence (including ```<bos>```, ```<eos>```) and choose correct rnn_output at [here](https://github.com/dongjun-Lee/birnn-language-model-tf/blob/master/model/bi_rnn_lm.py#L35-L36).

**Comment by dongjun-Lee** [comment]
@tsungruihon Sorry for that :(
This figure might help you to understand my implementation.
![image](https://user-images.githubusercontent.com/6512394/48831973-68ef9880-edbb-11e8-8fd5-f78162d15a5b.png)

**Comment by OswaldoBornemann** [comment]
@dongjun-Lee Then why in the `rnn_lm.py` , the `self.lm_input` is `self.x[:, :-1]`, let me make a example:

if `x` is `[1, 3, 4, 5, 2, 0]` then 

```
self.lm_input is [1,3, 4, 5, 2]
self.lm_output is [3, 4, 5, 2]
```
But in my world, i think the `self.lm_input` and `self.lm_output` should be below:

```
self.lm_input is [1,3, 4, 5]
self.lm_output is [3, 4, 5, 2]
```


if `x` is `[1, 3, 4, 5, 5, 2]` then 

```
self.lm_input is [1,3, 4, 5]
self.lm_output is [3, 4, 5, 2]
```
which is correct in my opinion.

**Comment by OswaldoBornemann** [comment]
@dongjun-Lee Can you explain more about the `self.lm_output` in `bi_rnn_lm.py` that why the `self.lm_output` should be `self.x[:, 1:-1]` ?

**Comment by dongjun-Lee** [comment]
@tsungruihon In ```bi_rnn_lm.py```, I added ```<bos>``` and ```<eos>``` to simplify the implementation. It uses full sentence including ```<bos>``` and ```<eos>``` as the RNN input but not using whole RNN output.
![image](https://user-images.githubusercontent.com/6512394/48833374-8b36e580-edbe-11e8-9f0e-e75c2c46dc4c.png)
This figure shows how my implementation works. We use [BOS, how, are, you, EOS] as an input sequence for RNN BUT NOT using whole RNN output. As [here](https://github.com/dongjun-Lee/birnn-language-model-tf/blob/master/model/bi_rnn_lm.py#L35-L36), real input and output (to calculate loss) for forward RNN is [BOS, how, are] -> [how, are, you] and for backward RNN [EOS, you, are] -> [you, are, how].

**Comment by OswaldoBornemann** [comment]
@dongjun-Lee  Have you tried to implement it to large corpus ?  like a billion

**Comment by dongjun-Lee** [comment]
No I haven't. I've trained the model for only ptb dataset.

**Comment by OswaldoBornemann** [comment]
Now i tried to implement the [adaptive softmax](https://github.com/yangsaiyong/tf-adaptive-softmax-lstm-lm/blob/master/train_lm.py) to your code and would like to see whether `adaptive softmax` could speed up bi-lstm.

**Comment by dongjun-Lee** [comment]
Sounds nice!

**Comment by OswaldoBornemann** [comment]
@dongjun-Lee Thank you a lot. Really appreciated！！！
